website/source/docs/internals/internal-plugins.html.md

```diff
@@ -0,0 +1,75 @@
+
+However, when you upgrade you will need to manually delete old plugins from disk. You can do this via something like this, depending on where you installed `terraform`:
+
+	rm /usr/local/bin/terraform-*
+
+If you don't do this you will see an error message like the following:
+
+```
+[WARN] /usr/local/bin/terraform-provisioner-file overrides an internal plugin for file-provisioner.
+  If you did not expect to see this message you will need to remove the old plugin.
+  See https://www.terraform.io/docs/internals/plugins.html
+Error configuring: 2 error(s) occurred:
+
+* Unrecognized remote plugin message: 2|unix|/var/folders/pj/66q7ztvd17v_vgfg8c99gm1m0000gn/T/tf-plugin604337945
+
+This usually means that the plugin is either invalid or simply
+needs to be recompiled to support the latest protocol.
+* Unrecognized remote plugin message: 2|unix|/var/folders/pj/66q7ztvd17v_vgfg8c99gm1m0000gn/T/tf-plugin647987867
+
+This usually means that the plugin is either invalid or simply
+needs to be recompiled to support the latest protocol.
+```
+
+## Why Does This Happen?
+
+In previous versions of Terraform all of the plugins were included in a zip file. For example, when you upgraded from 0.6.12 to 0.6.15, the newer version of each plugin file would have replaced the older one on disk, and you would have ended up with the latest version of each plugin.
+
+Going forward there is only one file in the distrubtion so you will need to perform a one-time cleanup when upgrading from Terraform < 0.7 to Terraform 0.7 or higher.
+
+If you're curious about the low-level details, keep reading!
+
+## Go Plugin Architecture
+
+Terraform is written in the Go programming language. One of Go's interesting properties is that it produces statically-compiled binaries. This means that it does not need to find libraries on your computer to run, and in general only needs to be compatible with your operating system (to make system calls) and with your CPU architecture (so the assembly instructions match the CPU you're running on).
+
+Another property of Go is that it does not support dynamic libraries. It _only_ supports static binaries. This is part of Go's overall design and is the reason why it produces statically-compiled binaries in the first place -- once you have a Go binary for your platform it should _Just Work_.
+
+In other languages, plugins are built using dynamic libraries. Since this is not an option for us in Go we use a network RPC interface instead. This means that each plugin is an idependent program, and instead of communicating via shared memory, the main process communicates with the plugin process over HTTP. When you start Terraform, it identifies the plugin you want to use, finds it on disk, runs the other binary, and does some handshaking to make sure they can talk to eachother (the error you may see after upgrading is a handshake failure in the RPC code).
```

Comment:
Derp. Thanks for pointing that out. I got smart and pressed spellcheck in Sublime. :-X